fish-rust/src/common.rs

```diff
@@ -1074,7 +1074,8 @@ pub static EMPTY_STRING_LIST: Vec<WString> = vec![];
 
 /// A function type to check for cancellation.
 /// \return true if execution should cancel.
-pub type CancelChecker = dyn Fn() -> bool;
+/// This needs a lifetime to allow closures that capture for less than static
+pub type CancelChecker<'closure> = dyn FnMut() -> bool + 'closure;
```

Comment:
This might actually disagree with what `OperationContext` want — in that case these should rather be separate types as `FnMut` is more general than `Fn`

Reply:
I'm not sure we want to change all references to `&mut` just for this type; its implementations don't mutate anything (though you're right they could in theory).

Reply:
(references of OperationContext I mean)

Reply:
We just need to take it by value here in `Wildcard`, see this playground https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ce59c7510e0307fa51c18c28bc795e74

Reply:
My suggestion: Take it by value, get rid of the type alias (there are a lot of type-aliases that I honestly think just obscure the code, this is one of them)

Reply:
I don't know about removing the type entirely, it does serve as immediate documentation in function signatures. How about making it a transparent newtype (`pub struct CancelChecker<'closure>(pub dyn FnMut() -> bool + 'closure);`) instead? Maybe with a `.check()` method if you don't want to access `.0` all the time.

Reply:
I don't think that is a valuable abstraction _for fish_. We already have `OperationContext::check_cancel`.

Reply:
Right, in the C++ version, wildcard expander also took cancel checker by value, so our default should be to match that.
In general there are lots of things to clean up, would be nice if someone could tackle that in future.

Reply:
Just finish the port first, then cleaning stuff up will be much easier :)

Reply:
Done https://github.com/fish-shell/fish-shell/pull/9916/commits/54437d70314a429af0afbbfa7d9c535cef3ee011, I had to rebase to make CI pass as well

Reply:
does this mean that OperationContext references need to be `mut`? I haven't looked into why we actually need FnMut

Reply:
No, as the playground showed you can use `&dyn Fn() -> bool` where the type is `impl FnMut() -> bool + 'closure`.

The reason I changed it to `FnMut` was because I was trying to add FFI-bindings, and that ended up something like:

```rust
// this is pseudo-code
fn use_opaque(x: &mut std::function-ish) -> bool {
    call(x)
}

let opaque_cpp_cancel_checker: std::function-ish = X;
let cancel_checker: impl FnMut() -> bool = || {
    // we capture `x` here for the lifetime of the closure, which is 'closure
    use_opaque(&mut opaque_cpp_cancel_checker)
};
```

And I got compiler errors due to `FnMut`, and my reaction was: hmm, it makes sense to allow `FnMut`, let me change that

Reply:
great